src/ripple/app/tx/impl/InvariantCheck.cpp

```diff
@@ -358,6 +358,101 @@ AccountRootsNotDeleted::finalize(
 
 //------------------------------------------------------------------------------
 
+void
+AccountRootsDeletedClean::visitEntry(
+    bool isDelete,
+    std::shared_ptr<SLE const> const& before,
+    std::shared_ptr<SLE const> const&)
+{
+    if (isDelete && before && before->getType() == ltACCOUNT_ROOT)
+        accountsDeleted_.emplace_back(before);
+}
+
+bool
+AccountRootsDeletedClean::finalize(
+    STTx const& tx,
+    TER const result,
+    XRPAmount const,
+    ReadView const& view,
+    beast::Journal const& j)
+{
+    // This list should include all of the keylet functions that take a single
+    // AccountID parameter, except nftpage_max, because it is exercised below
+    // using the NFT page lookup logic. (nftpage_min is included because it
+    // should _never_ be present.)
+    using fType = std::function<Keylet(AccountID const& id)>;
+    static std::array<fType, 4> const keyletfuncs{
+        &keylet::account,
+        &keylet::ownerDir,
```

Comment:
I'm not familiar with how the owner directory structure works. Does not having the first owner directory page mean that there aren't subsequent pages still in existence?

Reply:
It should. More to the point, though, the directory root is the only page I can look up with only the `AccountID`. I want to add more extensive directory checking down the line.

Background: The directory structure is basically a doubly-linked list. The root has index 0, and contains optional fields for `IndexNext` and `IndexPrevious`. If there is more than one page, those fields can be used in `page()` to find those pages. The page pointed to by `IndexNext` will have an `IndexPrevious` field of 0, and vice versa.